src/django/api/models.py

```diff
@@ -2112,3 +2140,64 @@ def __str__(self):
         return (
             'Contributor ID: {contributor_id} - ' +
             'Column Name: {column_name}').format(**self.__dict__)
+
+
+@transaction.atomic
+def index_facilities(facility_ids=[]):
+    # If passed an empty array, create or update all existing facilities
+    if len(facility_ids) == 0:
+        print('Indexing all facilities...')
+        facility_ids = Facility.objects.all().values_list('id', flat=True)
+
+    contrib_type = 'facility_list_item__source__contributor__contrib_type'
+    contributor = 'facility_list_item__source__contributor'
+    list = 'facility_list_item__source__facility_list_id'
+    filter = (Q(is_active=True)
+              & Q(status__in=[FacilityMatch.AUTOMATIC,
+                              FacilityMatch.CONFIRMED,
+                              FacilityMatch.MERGED])
+              & Q(facility_list_item__source__is_active=True)
+              & Q(facility_list_item__source__is_public=True))
+
+    # Create a list of dictionaries in the structure of FacilityIndexes
+    data = FacilityMatch.objects \
+                        .filter(facility_id__in=facility_ids) \
+                        .annotate(name=F('facility__name'),
+                                  country_code=F('facility__country_code'),
+                                  location=F('facility__location')) \
+                        .values('name', 'country_code', 'location') \
+                        .annotate(contrib_types=ArrayAgg(
+                                    contrib_type,
+                                    filter=filter),
+                                  contributors=ArrayAgg(
+                                    contributor,
+                                    filter=filter),
+                                  lists=ArrayAgg(
+                                    list,
+                                    filter=filter),
+                                  id=F('facility_id'),
+                                  ppe=Concat('facility__ppe_product_types',
+                                             'facility__ppe_contact_phone',
+                                             'facility__ppe_contact_email',
+                                             'facility__ppe_website',
+                                             output_field=CharField()))
```

Comment:
This is a very clever use of ORM features to handle both finding data and formatting it for insert in one pass. Well done. 👍